### PR TITLE
add nc_off tests in AES self tests for CTR mode

### PR DESCRIPTION
**Pull request by RonEld**
## Description
Add in the AES selftest tests for different `nc_off`

## Status
**READY**

## Requires Backporting
NO  

## Todos
- [x] Tests
- [ ] Documentation
- [x] Changelog updated
- [ ] Backported

## Steps to test or reproduce
run the `./test_suite_aes.rest` test suite

Note: should be merged together with #1178

### CODE REVIEW

File: library/aes.c [hunk 1]
```diff
@@ -1446,44 +1446,66 @@ int mbedtls_aes_self_test( int verbose )
...
+            while ( tst_offset < 16 )
             {
-                if( verbose != 0 )
-                    mbedtls_printf( "failed\n" );
-
-                ret = 1;
-                goto exit;
+                memcpy( buf, aes_test_ctr_ct[u], len );
+                memcpy( nonce_counter, aes_test_ctr_nonce_counter[u], 16 );
+                offset = 0;
```
Comment (mazimkhan) [inline library/aes.c]: Shouldn't ```offset``` i.e. ```nc_off``` be set outside the loop? As iterations intend to resume ciphering in stream ```stream_block```?

Comment (RonEld) [reply library/aes.c]: The `nc_off` is an IN\OUT parameter, and gets modified to the offset of the last block.  I want every iteration to start from the beginning of the stream ( offset 0 )
Perhaps the naming is not good.

**Comment by RonEld** [comment]
I improved the test, by trying to test all possible lengths, covering all possible corner cases:
resuming on first AES block, resuming on middle AES block,
resuming on last full AES block and resuming on last partial AES block

File: library/aes.c [hunk 2]
```diff
@@ -1446,44 +1446,66 @@ int mbedtls_aes_self_test( int verbose )
...
         {
             len = aes_test_ctr_len[u];
-            memcpy( buf, aes_test_ctr_ct[u], len );
-
-            mbedtls_aes_crypt_ctr( &ctx, len, &offset, nonce_counter, stream_block,
-                           buf, buf );
-
-            if( memcmp( buf, aes_test_ctr_pt[u], len ) != 0 )
+            size_to_cipher = 0;
+            while ( size_to_cipher < len )
```
Comment (mazimkhan) [inline library/aes.c]: Cosmetic: A for loop will be more neat.

File: library/aes.c [hunk 3]
```diff
@@ -1446,44 +1446,79 @@ int mbedtls_aes_self_test( int verbose )
...
-                if( verbose != 0 )
-                    mbedtls_printf( "failed\n" );
-
-                ret = 1;
-                goto exit;
+                memcpy( buf, aes_test_ctr_ct[u], len );
+                memcpy( nonce_counter, aes_test_ctr_nonce_counter[u], 16 );
+                offset = 0;
+
+                /* cipher all the blocks until the last one, and add the offset*/
```
Comment (mazimkhan) [inline library/aes.c]: This comment is not true. For ```length == 32```, iterations ```size_to_cipher < 16``` only first block is ciphered. It may be said as "beginning block(s)" and "remaining block(s)" for next call.

Comment (RonEld) [reply library/aes.c]: Thanks, I didn't change the comment when I updated the test

**Comment by mazimkhan** [comment]
Good. Adding this check would save debugging time on failure.

**Comment by mazimkhan** [comment]
LGTM

**Comment by RonEld** [comment]
I created an issue for this, for tracking, in #1177

**Comment by gilles-peskine-arm** [comment]
Note that there is a conflict that needs to be resolved.

**Comment by RonEld** [comment]
I rebased and fixed the conflict

File: ChangeLog [hunk 4]
```diff
@@ -378,6 +378,9 @@ Changes
    * Add explicit warnings for the use of MD2, MD4, MD5, SHA-1, DES and ARC4
      throughout the library.
 
+Changes
+   * Add tests for different nc_off in AES CTR mode, in the AES selftests
```
Comment (Patater) [inline ChangeLog]: Do we really want a ChangeLog entry for adding tests?

File: library/aes.c [hunk 5]
```diff
@@ -1523,35 +1523,90 @@ int mbedtls_aes_self_test( int verbose )
...
 
         len = aes_test_ctr_len[u];
 
         if( mode == MBEDTLS_AES_DECRYPT )
         {
-            memcpy( buf, aes_test_ctr_ct[u], len );
-            aes_tests = aes_test_ctr_pt[u];
+            len = aes_test_ctr_len[u];
+
+            for(size_to_cipher = 0; size_to_cipher < len; size_to_cipher++ )
```
Comment (Patater) [inline library/aes.c]: Style: Add a space after `for(`

File: library/aes.c [hunk 6]
```diff
@@ -1523,35 +1523,90 @@ int mbedtls_aes_self_test( int verbose )
...
+            {
+                memcpy( buf, aes_test_ctr_ct[u], len );
+                memcpy( nonce_counter, aes_test_ctr_nonce_counter[u], 16 );
+                offset = 0;
+
+                /* cipher the first bytes with length size_to_cipher */
+                ret = mbedtls_aes_crypt_ctr( &ctx,  size_to_cipher, &offset, nonce_counter, stream_block,
+                                       buf, buf );
```
Comment (Patater) [inline library/aes.c]: Style: Align the `buf, buf` line with the parameters above.

File: library/aes.c [hunk 7]
```diff
@@ -1523,35 +1523,90 @@ int mbedtls_aes_self_test( int verbose )
...
+                {
+                    if( verbose != 0 )
+                        mbedtls_printf( "failed: calculated offset is not as expected\n" );
+                    ret = 1;
+                    goto exit;
+                }
+
+                /* resume the cipher, from offset */
+                ret = mbedtls_aes_crypt_ctr( &ctx, len - size_to_cipher, &offset, nonce_counter, stream_block,
+                                       buf + size_to_cipher, buf  + size_to_cipher );
```
Comment (Patater) [inline library/aes.c]: Style: Please align the parameters.

File: library/aes.c [hunk 8]
```diff
@@ -1523,35 +1523,90 @@ int mbedtls_aes_self_test( int verbose )
...
+                    goto exit;
+                if( memcmp( buf, aes_test_ctr_pt[u], len ) != 0 )
+                {
+                    if( verbose != 0 )
+                        mbedtls_printf( "failed with cipher size %zu\n", size_to_cipher );
+
+                    ret = 1;
+                    goto exit;
+                }
+            }
```
Comment (Patater) [inline library/aes.c]: Please refactor the `if( mode == MBEDTLS_AES_DECRYPT )` blocks, as there is a lot of duplicated code and the differences between them should be clearer.

File: library/aes.c [hunk 9]
```diff
@@ -1523,35 +1523,90 @@ int mbedtls_aes_self_test( int verbose )
...
+                if( offset != ( size_to_cipher % 16 ) )
+                {
+                    if( verbose != 0 )
+                        mbedtls_printf( "failed: calculated offset is not as expected\n" );
+                    ret = 1;
+                    goto exit;
+                }
+
+                /* resume the cipher, from offset */
+                ret = mbedtls_aes_crypt_ctr( &ctx, len- size_to_cipher, &offset, nonce_counter, stream_block,
```
Comment (Patater) [inline library/aes.c]: Style: Add spaces around binary operators like `-`.

**Comment by RonEld** [comment]
Now that I think about it, I am not sure this PR is needed. I mean, these tests are needed, but I think they are more suitable for our tests suites, rather than the self tests. The self tests are more of a sanity check.
Comments?

**Comment by sbutcher-arm** [comment]
There's no need to have `nc_off` tests in a self-test.

To be clear - self-tests are supposed to be functional tests that can be executed as start-up or diagnostic tests on the final production device, as a way of confirming that the device is functioning correctly. Sometimes these tests may be executed to confirm a device hasn't been compromised by an attack.

The `nc_off` test is a design test to validate the implementation - and specifically - this is a robustness provision. There's no reason, nor advantage to execute it on a final device, so I don't think it belongs in a self-test.

**Comment by RonEld** [comment]
Note: We have #1085 to track missing ctr tests

**Comment by RonEld** [comment]
closing this PR, until we have ctr tests in our test suite, and we integrate these tests there
